File: backend/routers/rebalance.py
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
import pandas as pd

from ..utils.cache import portfolio_cache, load_portfolio
from ..utils.cache import sector_info_cache

router = APIRouter()


def _get_last_price(stock_obj) -> float:
    close = stock_obj.data["Close"]
    val = close.iloc[-1]
    return float(val.iloc[0]) if isinstance(val, pd.Series) else float(val)


class TradeRequest(BaseModel):
    target_weights: dict[str, float]  # ticker -> weight (must sum to ~1.0)
    total_budget: float = 0.0         # optional override; defaults to current portfolio value


@router.get("/current_allocation")
def get_current_allocation():
    """Return current holdings with weights, prices, and sector labels."""
    if not portfolio_cache:
        return {"status": "loading", "message": "Portfolio data not yet loaded."}

    holdings = load_portfolio()
    ticker_meta = {h["ticker"]: h for h in holdings}

    rows = []
    total_value = 0.0

    for ticker, stock_obj in portfolio_cache.items():
        try:
            price = _get_last_price(stock_obj)
            value = price * stock_obj.shares
            total_value += value
            rows.append({"ticker": ticker, "price": price, "value": value})
        except Exception:
            continue

    if total_value == 0:
        return {"error": "Could not calculate portfolio value."}

    assets = []
    sector_totals: dict[str, float] = {}

    for row in rows:
        ticker = row["ticker"]
        info = sector_info_cache.get(ticker, {})
        sector = info.get("sector") or "Unknown"
        weight = round(row["value"] / total_value, 6)
        sector_totals[sector] = sector_totals.get(sector, 0) + weight
        assets.append({
            "ticker": ticker,
            "shares": portfolio_cache[ticker].shares,
            "price": round(row["price"], 2),
            "value": round(row["value"], 2),
            "current_weight": weight,
            "sector": sector,
            "description": ticker_meta.get(ticker, {}).get("description", ticker),
        })

    assets.sort(key=lambda x: x["value"], reverse=True)

    return {
        "assets": assets,
        "total_value": round(total_value, 2),
        "sector_weights": {k: round(v, 4) for k, v in sorted(sector_totals.items(), key=lambda x: x[1], reverse=True)},
    }


@router.post("/calculate_trades")
def calculate_trades(req: TradeRequest):
    """
    Given target weights (must sum to 1.0) and optionally a total budget,
    return the list of buy/sell trades needed to rebalance.
    """
    if not portfolio_cache:
        raise HTTPException(status_code=503, detail="Portfolio data not yet loaded.")

    weights = req.target_weights
    weight_sum = sum(weights.values())
    if abs(weight_sum - 1.0) > 0.01:
        raise HTTPException(
            status_code=400,
            detail=f"Target weights must sum to 1.0 (got {weight_sum:.4f})."
        )

    # Build current state
    rows = {}
    total_value = 0.0
    for ticker, stock_obj in portfolio_cache.items():
        try:
            close = stock_obj.data["Close"]
            val = close.iloc[-1]
            price = float(val.iloc[0]) if isinstance(val, pd.Series) else float(val)
            value = price * stock_obj.shares
            total_value += value
            rows[ticker] = {"price": price, "value": value, "shares": stock_obj.shares}
        except Exception:
            continue

    budget = req.total_budget if req.total_budget > 0 else total_value

    trades = []
    for ticker, target_w in weights.items():
        current_value = rows.get(ticker, {}).get("value", 0.0)
        current_w = current_value / total_value if total_value > 0 else 0.0
        price = rows.get(ticker, {}).get("price")

        if price is None or price == 0:
            continue

        target_value = target_w * budget
        delta_value = target_value - current_value
        delta_shares = int(delta_value / price)  # whole shares only

        action = "BUY" if delta_shares > 0 else ("SELL" if delta_shares < 0 else "HOLD")

        trades.append({
            "ticker": ticker,
            "action": action,
            "shares_delta": delta_shares,
            "current_shares": rows.get(ticker, {}).get("shares", 0),
            "target_shares": rows.get(ticker, {}).get("shares", 0) + delta_shares,
            "price": round(price, 2),
            "trade_value": round(abs(delta_shares * price), 2),
            "current_weight": round(current_w, 4),
            "target_weight": round(target_w, 4),
            "drift": round(current_w - target_w, 4),
        })

    trades.sort(key=lambda x: abs(x["drift"]), reverse=True)

    buys = sum(t["trade_value"] for t in trades if t["action"] == "BUY")
    sells = sum(t["trade_value"] for t in trades if t["action"] == "SELL")
    turnover_pct = round((buys + sells) / 2 / budget * 100, 2) if budget else 0

    return {
        "trades": trades,
        "summary": {
            "total_buys": round(buys, 2),
            "total_sells": round(sells, 2),
            "turnover_pct": turnover_pct,
            "portfolio_value": round(total_value, 2),
            "budget": round(budget, 2),
        },
    }


@router.get("/equal_weight_target")
def equal_weight_target():
    """Return equal-weight target allocation for all portfolio tickers."""
    holdings = load_portfolio()
    n = len(holdings)
    w = round(1.0 / n, 6)
    return {h["ticker"]: w for h in holdings}

File: frontend/js/rebalance.js
import { get, post } from "./api.js";

const currency = new Intl.NumberFormat("en-US", { style: "currency", currency: "USD" });
let _currentAllocation = null;

export function init() {
    loadAllocation();
    document.getElementById("btn-equal-weight").addEventListener("click", applyEqualWeight);
    document.getElementById("btn-calculate-trades").addEventListener("click", calculateTrades);
}

async function loadAllocation() {
    setStatus("Loading current allocation...");
    try {
        const data = await get("/rebalance/current_allocation");
        if (data.status === "loading") {
            setStatus(data.message);
            setTimeout(loadAllocation, 4000);
            return;
        }
        _currentAllocation = data;
        renderAllocationTable(data);
        setStatus("");
    } catch (err) {
        setStatus("Error: " + err.message);
    }
}

function renderAllocationTable(data) {
    const tbody = document.getElementById("rebalance-holdings-body");
    tbody.innerHTML = "";
    data.assets.forEach(asset => {
        const tr = document.createElement("tr");
        tr.innerHTML = `
            <td><strong>${asset.ticker}</strong></td>
            <td>${asset.shares.toLocaleString()}</td>
            <td>${currency.format(asset.price)}</td>
            <td>${currency.format(asset.value)}</td>
            <td>${(asset.current_weight * 100).toFixed(2)}%</td>
            <td>${asset.sector}</td>
            <td><input type="number" class="weight-input" data-ticker="${asset.ticker}"
                 value="${(asset.current_weight * 100).toFixed(2)}"
                 min="0" max="100" step="0.01" style="width:70px"></td>
        `;
        tbody.appendChild(tr);
    });

    document.getElementById("rebalance-total-value").innerText =
        `Current portfolio value: ${currency.format(data.total_value)}`;
}

async function applyEqualWeight() {
    try {
        const weights = await get("/rebalance/equal_weight_target");
        document.querySelectorAll(".weight-input").forEach(input => {
            const ticker = input.dataset.ticker;
            if (ticker in weights) {
                input.value = (weights[ticker] * 100).toFixed(2);
            }
        });
        setStatus("Equal weights applied. Click 'Calculate Trades' to see required trades.");
    } catch (err) {
        setStatus("Error: " + err.message);
    }
}

async function calculateTrades() {
    const inputs = document.querySelectorAll(".weight-input");
    const targetWeights = {};
    let total = 0;

    inputs.forEach(input => {
        const w = parseFloat(input.value) / 100;
        targetWeights[input.dataset.ticker] = w;
        total += w;
    });

    if (Math.abs(total - 1.0) > 0.01) {
        setStatus(`Error: Weights sum to ${(total * 100).toFixed(2)}% — must equal 100%.`);
        return;
    }

    setStatus("Calculating trades...");
    try {
        const result = await post("/rebalance/calculate_trades", {
            target_weights: targetWeights,
            total_budget: 0,
        });
        renderTrades(result);
        setStatus("");
    } catch (err) {
        setStatus("Error: " + err.message);
    }
}

function renderTrades(result) {
    const tbody = document.getElementById("trades-body");
    tbody.innerHTML = "";

    result.trades.forEach(t => {
        if (t.action === "HOLD") return;
        const tr = document.createElement("tr");
        tr.className = t.action === "BUY" ? "trade-buy" : "trade-sell";
        tr.innerHTML = `
            <td><strong>${t.ticker}</strong></td>
            <td class="${t.action === "BUY" ? "text-green" : "text-red"}">${t.action}</td>
            <td>${t.shares_delta > 0 ? "+" : ""}${t.shares_delta}</td>
            <td>${currency.format(t.price)}</td>
            <td>${currency.format(t.trade_value)}</td>
            <td>${(t.current_weight * 100).toFixed(2)}%</td>
            <td>${(t.target_weight * 100).toFixed(2)}%</td>
            <td>${(t.drift * 100).toFixed(2)}%</td>
        `;
        tbody.appendChild(tr);
    });

    const s = result.summary;
    document.getElementById("trades-summary").innerHTML = `
        <span>Total Buys: <strong>${currency.format(s.total_buys)}</strong></span>
        &nbsp;|&nbsp;
        <span>Total Sells: <strong>${currency.format(s.total_sells)}</strong></span>
        &nbsp;|&nbsp;
        <span>Turnover: <strong>${s.turnover_pct}%</strong></span>
    `;

    // Drift bar chart
    const drifts = result.trades.filter(t => t.action !== "HOLD");
    const colors = drifts.map(t => t.drift > 0 ? "#d9534f" : "#5cb85c");
    const fig = {
        data: [{
            type: "bar",
            x: drifts.map(t => t.ticker),
            y: drifts.map(t => +(t.drift * 100).toFixed(2)),
            marker: { color: colors },
            hovertemplate: "%{x}: %{y:.2f}% drift<extra></extra>",
        }],
        layout: {
            title: "Portfolio Drift (Current − Target Weight %)",
            xaxis: { tickangle: -45 },
            yaxis: { title: "Drift (%)" },
            shapes: [{ type: "line", x0: 0, x1: 1, xref: "paper", y0: 0, y1: 0, line: { color: "#333", dash: "dot" } }],
            margin: { b: 100 },
        }
    };
    Plotly.react("drift-chart", fig.data, fig.layout, { responsive: true });
}

function setStatus(msg) {
    const el = document.getElementById("rebalance-status");
    if (el) el.innerText = msg;
}
